jdb/assignment2/tree_no_parent.py

```diff
@@ -0,0 +1,97 @@
+import unittest
+from typing import Generator, List, Tuple, TypeVar
+
+T = TypeVar('T')
+TAncestors = Tuple[T]
+_StackOfT = List[T]
+
+class Node:
+    def __init__(self, value: T, left = None, right = None):
+        self.left = Node(*left) if left else None
```

Comment:
I'm not sure about what "*" is meaning there. Why not to use just Node(left) etc?

Reply:
The start, applied to a list of a tuple "unpacks" the element of the list or tuple each in the corresponding arguments. See this example:

>>> args = '1st', '2nd', '3rd'
>>> def f(a,b,c):
...   print('All good')
... 
>>> f(*args)
All good
>>> f(args)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: f() takes exactly 3 arguments (1 given)

The error message is pretty clear: args was passed as the first argument it became 'a', and b and c are missing.

The advantage of this notation is to separate the time where you can create, manipulate, transfer the arguments of a function, and in a second, later time, you can use arguments and apply the function on them.

For example, in the unit tests, the recursive tuple of tuple stores the arguments later passed to the Node constructor.

Reply:
Yes, thanks